src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs

```diff
@@ -615,6 +618,84 @@ internal static void RegisterMemoryLoadChangeNotification(float lowMemoryPercent
             }
         }
 
+        [Flags]
+        internal enum FinalizerActionKind
+        {
+            None = 0,
+            ExecuteNoGCCallback = 1,
+            DetachNoGCCallback = 2,
+        }
+
+        internal enum EnableNoGCRegionCallbackStatus
+        {
+            Succeed,
+            NotStarted,
+            InsufficientBudget,
+        }
+
+        private static Action? s_NoGCCallback;
+
+        [UnmanagedCallersOnly]
+        private static unsafe void FinalizerAction(FinalizerActionKind finalizerActionKind)
+        {
+            Action? callback = s_NoGCCallback;
+            if ((finalizerActionKind & FinalizerActionKind.DetachNoGCCallback) != 0)
+            {
+                s_NoGCCallback = null;
+            }
+            if ((finalizerActionKind & FinalizerActionKind.ExecuteNoGCCallback) != 0)
+            {
+                callback?.Invoke();
+            }
+        }
+
+        /// <summary>
+        /// Register a callback to be invoked when we allocated a certain amount of memory in the no GC region.
+        /// <param name="totalSize">The total size of the no GC region. Must be a number > 0 or an ArgumentOutOfRangeException will be thrown.</param>
+        /// <param name="callback">The callback to be executed when we allocated a certain amount of memory in the no GC region..</param>
+        /// <exception cref="System.ArgumentOutOfRangeException"> The <paramref name="totalSize"/> argument is less than or equal to 0.</exception>
+        /// <exception cref="System.ArgumentNullException">The <paramref name="callback"/> argument is null.</exception>
+        /// <exception cref="InvalidOperationException"><para>The GC is not currently under a NoGC region.</para>
+        /// <para>-or-</para>
+        /// <para>Another callback is already registered.</para>
+        /// <para>-or-</para>
+        /// <para>The <paramref name="totalSize"/> exceeds the size of the No GC region.</para>
+        /// <para>-or-</para>
+        /// <para>We failed to withheld memory for the callback before of already made allocation.</para>
+        /// </exception>
+        /// </summary>
+        public static unsafe void RegisterNoGCRegionCallback(long totalSize, Action callback)
+        {
+            ArgumentOutOfRangeException.ThrowIfNegativeOrZero(totalSize);
+            ArgumentNullException.ThrowIfNull(callback);
+
+            if (Interlocked.CompareExchange(ref s_NoGCCallback, callback, null) != null)
+            {
+                throw new InvalidOperationException(SR.InvalidOperationException_NoGCRegionCallbackAlreadyRegistered);
+            }
+
+            _RegisterFinalizerAction(&FinalizerAction);
```

Comment:
`_RegisterFinalizerAction` and `_EnableNoGCRegionCallback` can be merged into a single call.